File: Controllers/DepartmentController.cs
using DMS.Models;
using Newtonsoft.Json;
using Microsoft.AspNetCore.Http.HttpResults;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Identity.Client;
using System;
using System.Diagnostics;
using Microsoft.EntityFrameworkCore.SqlServer;

namespace DMS.Controllers
{
    public class DepartmentController : Controller
    {
        public const string SessionKeyName = "username";
        private readonly ILogger<DepartmentController> _logger;
        private readonly IHttpContextAccessor contxt;

        private DMS_DbContext dataContext { get; set; }

        public DepartmentController(ILogger<DepartmentController> logger, DMS_DbContext _dataContext, IHttpContextAccessor httpContextAccessor)
        {
            _logger = logger;
            dataContext = _dataContext;
            contxt = httpContextAccessor;
        }

        public IActionResult Create()
        {
            TempData["username"] = HttpContext.Session.GetString(SessionKeyName);
            return View();
        }

        [HttpPost]
        [ValidateAntiForgeryToken]
        public IActionResult Create(DepartmentModel department)
        {
            // Validate if the name and age exist in the database
            bool nameExist = dataContext.Department.Any(p => p.Name == department.Name);
            bool codeExist = dataContext.Department.Any(p => p.Code == department.Code);

            if (nameExist)
            {
                ViewBag.NameAlert = "<span class='text-danger'>This Name already Exist</span>";
                return View(department);
            }
            if (department.Name == null)
            {
                ViewBag.NameAlert = "<span class='text-danger'>Please Input a Name</span>";
                return View(department);
            }
            if (codeExist)
            {
                ViewBag.CodeAlert = "<span class='text-danger'>This Code already Exist</span>";
                return View(department);
            }
            if (department.Code == null)
            {
                ViewBag.CodeAlert = "<span class='text-danger'>Please Input a Code</span>";
                return View(department);
            }

            department.CreatedDate = DateTime.Now;
            department.Active = true;
            department.CreatedBy = HttpContext.Session.GetString(SessionKeyName);
            dataContext.Department.Add(department);
            dataContext.SaveChanges();
            return RedirectToAction("Index");
        }

        [HttpPost]
        [ValidateAntiForgeryToken]
        public IActionResult Edit(DepartmentModel department)
        {
            var name = (string)TempData["Name"];
            var code = (string)TempData["Code"];

            // Validate if the name and age exist in the database
            bool nameExist = dataContext.Department.Any(p => p.Name == department.Name && p.Name != name);
            bool codeExist = dataContext.Department.Any(p => p.Code == department.Code && p.Code != code);

            var active = (bool)TempData["Active"];
            var CreatedBy = (string)TempData["CreatedBy"];
            var CreatedDate = (DateTime)TempData["CreatedDate"];

            if (department.Name == name && department.Code == code)
            {
                ViewBag.CodeAlert = "<span class='text-danger'>You don't change any of the data!</span>";
                LoadDepartmentTemp(name, code, active, CreatedBy, CreatedDate);
                return View(department);
            }
            if (nameExist)
            {
                ViewBag.NameAlert = "<span class='text-danger'>This Name already Exist</span>";
                LoadDepartmentTemp(name, code, active, CreatedBy, CreatedDate);
                return View(department);
            }
            if (department.Name == null)
            {
                ViewBag.NameAlert = "<span class='text-danger'>Please Input a Name</span>";
                LoadDepartmentTemp(name, code, active, CreatedBy, CreatedDate);
                return View(department);
            }
            if (codeExist)
            {
                ViewBag.CodeAlert = "<span class='text-danger'>This Code already Exist</span>";
                LoadDepartmentTemp(name, code, active, CreatedBy, CreatedDate);
                return View(department);
            }
            if (department.Code == null)
            {
                ViewBag.CodeAlert = "<span class='text-danger'>Please Input a Code</span>";
                LoadDepartmentTemp(name, code, active, CreatedBy, CreatedDate);
                return View(department);
            }

            department.CreatedDate = CreatedDate;
            department.Active = active;
            department.CreatedBy = CreatedBy;
            department.ModifyDate = DateTime.Now;
            department.ModifyBy = HttpContext.Session.GetString(SessionKeyName);
            dataContext.Department.Update(department);
            dataContext.SaveChanges();
            return RedirectToAction("Index");
        }

        public void LoadDepartmentTemp(string name, string code, bool active, string CreatedBy, DateTime CreatedDate)
        {
            TempData["Name"] = name;
            TempData["Code"] = code;
            TempData["Active"] = active;
            TempData["CreatedBy"] = CreatedBy;
            TempData["CreatedDate"] = CreatedDate;
        }

        public IActionResult Edit(int id)
        { 
            if (id == null || id == 0)
                return NotFound();
            
            var department = dataContext.Department.Find(id);

            if (department == null) 
                return NotFound();

            TempData["Name"] = department.Name;
            TempData["Code"] = department.Code;
            TempData["Active"] = department.Active;
            TempData["CreatedBy"] = department.CreatedBy;
            TempData["CreatedDate"] = department.CreatedDate;
            return View(department);
        }

        public IActionResult Delete(int id)
        {
            var obj = dataContext.Department.Find(id);
            if (obj == null)   
                return NotFound();

            dataContext.Department.Remove(obj);
            dataContext.SaveChanges();
            return RedirectToAction("Index");
        }

        public IActionResult Index()
        {
            // Retrieve the data from TempData
            var finalListJson = TempData["FinalList"] as string;

            if (finalListJson != null)
            {
                // Deserialize back to List<DepartmentModel>
                var finalList = JsonConvert.DeserializeObject<List<DepartmentModel>>(finalListJson);

                // Pass it to the view or process it as needed
                return View(finalList);
            }

            IEnumerable<DepartmentModel> DepartmentList = dataContext.Department;
            TempData["username"] = HttpContext.Session.GetString(SessionKeyName);
            return View(DepartmentList);
        }

        [HttpPost]
        public IActionResult Search(string? inputValue)
        {
            if (inputValue == null)
            {
                return RedirectToAction("Index");
            }

            List<DepartmentModel> nameDepartments = dataContext.Department
                                                            .Where(d => d.Name.Contains(inputValue))
                                                            .ToList();

            List<DepartmentModel> codeDepartments = dataContext.Department
                                                            .Where(d => d.Code.Contains(inputValue))
                                                            .ToList();

            var FinalList = nameDepartments.Concat(codeDepartments)
                                .Distinct()
                                .ToList();

            // Store in TempData (need to serialize to JSON or similar format)
            TempData["FinalList"] = JsonConvert.SerializeObject(FinalList);

            // Redirect to another action
            return RedirectToAction("Index");
        }

        [ResponseCache(Duration = 0, Location = ResponseCacheLocation.None, NoStore = true)]
        public IActionResult Error()
        {
            return View(new ErrorViewModel { RequestId = Activity.Current?.Id ?? HttpContext.TraceIdentifier });
        }
    }
}

File: Controllers/DepartmentController.js
const express = require('express');
const router = express.Router();
const db = require('../config/database'); 
const Department = db.Department
const SubDepartment = db.SubDepartment
const { Op } = require('sequelize');
const requireAuth = require('../middleware/requireAuth'); // assuming you have this

// GET: Create Department Form
router.get('/create', requireAuth, (req, res) => {
  res.render('department/create', { username: req.session.username });
});

// POST: Create Department
router.post('/create', requireAuth, async (req, res) => {
  const { Name, Code } = req.body;
  console.log("req.user",req.user)
  const CreatedBy = req.user.userName;

  try {
    if (!Name) return res.json( { NameAlert: "Please Input a Name", Code });
    if (!Code) return res.josn( { CodeAlert: "Please Input a Code", Name });

    const nameExist = await Department.findOne({ where: { Name } });
    if (nameExist) return res.json( { NameAlert: "This Name already exists", Code });

    const codeExist = await Department.findOne({ where: { Code } });
    if (codeExist) return res.json( { CodeAlert: "This Code already exists", Name });

    await Department.create({
      Name,
      Code,
      CreatedBy,
      CreatedDate: new Date(),
      Active: true,
    });

    res.json({
        status:true
    });
  } catch (err) {
    console.error(err);
    res.status(500).send('Server Error');
  }
});

// GET: Edit Form
router.get('/edit/:id', requireAuth, async (req, res) => {
  // const department = await Department.findByPk(req.params.id);
  const department = await Department.findOne({ where: { Id: req.params.id, Active: 1 } });
  if (!department) return res.status(404).send("Department not found");

  req.session.editData = department; // Store for comparison
  res.render('department/edit', department);
});

// POST: Edit Department
router.post('/edit', requireAuth, async (req, res) => {
  const { Id, Name, Code } = req.body;
//   const { Name: oldName, Code: oldCode, Active, CreatedBy, CreatedDate } = req.session.editData;
  const ModifyBy = req.user.username;

  try {
    if (!Name) return res.json( { NameAlert: "Please Input a Name", Code });
    if (!Code) return res.json( { CodeAlert: "Please Input a Code", Name });

    // if (Name === oldName && Code === oldCode) {
    //   return res.json('department/edit', {
    //     CodeAlert: "You didn't change any of the data!",
    //     Name, Code
    //   });
    // }

    const nameExist = await Department.findOne({ where: { Name, Id: { [Op.ne]: Id } } });
    if (nameExist) return res.json( { NameAlert: "This Name already exists", Code });

    const codeExist = await Department.findOne({ where: { Code, Id: { [Op.ne]: Id } } });
    if (codeExist) return res.json( { CodeAlert: "This Code already exists", Name });

    await Department.update({
      Name,
      Code,
      
      
      ModifyBy,
      ModifyDate: new Date(),
    }, {
      where: { Id }
    });

    res.json({status:true});
  } catch (err) {
    console.error(err);
    res.status(500).send('Server Error');
  }
});

// GET: Delete Department
// router.get('/delete/:id', requireAuth, async (req, res) => {
//   try {
//     await Department.destroy({ where: { Id: req.params.id } });
//     res.redirect('/department');
//   } catch (err) {
//     console.error(err);
//     res.status(500).send('Server Error');
//   }
// });
router.get('/delete/:id', requireAuth, async (req, res) => {
  const departmentId = req.params.id;

  try {
    // throw new Error("This is a test error");
    // Find all SubDepartments related to this Department
    // const subDepartments = await SubDepartment.findAll({
    //   where: { DepartmentID: departmentId }
    // });

    // Log them
    // console.error(`SubDepartments for Department ID ${departmentId}:`, subDepartments);
    
    // // Delete all related SubDepartments
    // await SubDepartment.destroy({
    //   where: { DepartmentID: departmentId }
    // });

    // // Delete the Department
    // await Department.destroy({
    //   where: { Id: departmentId }
    // });

    // Soft-delete: set Active = 0 for SubDepartments and Department inside a transaction
    await db.sequelize.transaction(async (t) => {
      await SubDepartment.update(
        { Active: 0 },
        { where: { DepartmentID: departmentId }, transaction: t }
      );

      await Department.update(
        { Active: 0 },
        { where: { Id: departmentId }, transaction: t }
      );
    });

    res.redirect('/department');
  } catch (err) {
    console.error(err);
    res.status(500).send(err.message || 'Server Error');
  }
});



// GET: List Departments
router.get('/', requireAuth, async (req, res) => {
  // Only list active departments and include only active subdepartments
  const list = await Department.findAll({
    where: { Active: 1 },
    include: [
      {
        model: db.SubDepartment,
        where: { Active: 1 },
        required: false // still return department if it has no active subdepartments
      }
    ]
  });
//   delete req.session.searchResults;
  res.json( { departments: list, username: req.user.username });
});

// POST: Search Department
router.post('/search', requireAuth, async (req, res) => {
  const { inputValue } = req.body;

  if (!inputValue) return res.json({
    status:"false"
  });

  const results = await Department.findAll({
    where: {
      [Op.and]: [
        { Active: 1 },
        {
          [Op.or]: [
            { Name: { [Op.like]: `%${inputValue}%` } },
            { Code: { [Op.like]: `%${inputValue}%` } }
          ]
        }
      ]
    }
  });

 res.json({
    status:true,
    results
 });
});

module.exports = router;
